src/NodaTime/Utility/TickArithmetic.cs

```diff
@@ -11,48 +12,90 @@ namespace NodaTime.Utility
-            }
-            else
-            {
-                // Make sure the multiplication doesn't overflow...
-                tickOfDay = ticks - (days + 1) * TicksPerDay + TicksPerDay;
+                    int days = (int)((ticks + 1) / TicksPerDay) - 1;
+                    // We need to be careful as ticks might be close to long.MinValue, at which point
+                    // days * TicksPerDay would overflow. We could validate that, but it only
+                    // saves two additions, and introduces a branch.
+                    tickOfDay = ticks - (days + 1) * TicksPerDay + TicksPerDay;
+                    return days;
+                }
             }
-            return days;
         }
 
-        internal static int TicksToDays(long ticks)
+        /// <summary>
+        /// Similar to <see cref="TicksToDaysAndTickOfDay(long, out long)"/> but
+        /// trusting that the input is non-negative, which can be proved in certain cases.
+        /// </summary>
+        /// <remarks>
+        /// Used by <see cref="LocalDateTime.FromDateTime(System.DateTime)"/> and 
+        /// <see cref="LocalDateTime.FromDateTime(System.DateTime, CalendarSystem)"/>.
+        /// </remarks>
+        internal static int NonNegativeTicksToDaysAndTickOfDay([Trusted] long ticks, out long tickOfDay)
         {
-            // First work out the number of days, always rounding down (so that ticks * TicksPerDay is always the
-            // start of the day).
-            // The shift approach here is equivalent to dividing by NodaConstants.TicksPerDay, but appears to be
-            // very significantly faster under the x64 JIT (and no slower under the x86 JIT).
-            // See http://stackoverflow.com/questions/22258070 for the inspiration.
-            return ticks >= 0 ? unchecked((int)((ticks >> 14) / 52734375L))
-                    // TODO: Optimize with shifting at some point. Note that this must *not* subtract from ticks,
-                    // as it could already be long.MinValue.
-                    : (int)((ticks + 1) / TicksPerDay) - 1;
+            unchecked
+            {
+                Preconditions.DebugCheckArgument(ticks >= 0, nameof(ticks), "Ticks must be non-negative");
+                int days = unchecked((int) ((ticks >> 14) / 52734375L));
+                tickOfDay = ticks - days * TicksPerDay;
+                return days;
+            }
         }
 
-        // TODO(2.0): Check whether this is still required. I don't believe we'll ever see
-        // a value with such a small number of ticks now, due to the decreased date range in 2.0.
-        internal static long DaysAndTickOfDayToTicks(int days, long tickOfDay) =>
+
+        /// <summary>
+        /// Cautiously computes a number of ticks from day/tick-of-day value. This may overflow,
+        /// but will only do so if it has to.
+        /// </summary>
+        internal static long DaysAndTickOfDayToTicksChecked(int days, long tickOfDay) =>
```

Comment:
Bounded sounds reasonable. I was originally thrown because the 'unsafe' version of `TicksToDaysAndTickOfDay()` still used `unchecked`...